File: Difficulty: Medium/Sort 0s, 1s and 2s/sort-0s-1s-and-2s.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;



// } Driver Code Ends
class Solution
{
    public:
    void sort012(int arr[], int n)
    {
        // code here 
        
        int i=0;
        int j=0;
        int k=n-1;
        while(j<=k){
         
            if(arr[j]==0)
            {
                swap(arr[i],arr[j]);
                i++;
                j++;
            }
           else if(arr[j]==2){
                swap(arr[j],arr[k]);
                k--;
            }
            else if(arr[j]==1){
               j++;
           }
        
    }
//     class Solution
// {
//     public:
//     void sort012(int a[], int n)
//     {
//         // code here 
//         int low = 0, mid = 0, high = n-1;
//         while(mid <= high){
//             if(a[mid] == 0){
//                 swap(a[low], a[mid]);
//                 low++;
//                 mid++;
//             }
//             else if(a[mid] == 2){
//                 swap(a[mid], a[high]);
//                 high--;
//             }
//             else{
//                 mid++;
//             }
//         }
//     }
    
};
    
  
    
    
};

//{ Driver Code Starts.
int main() {

    int t;
    cin >> t;

    while(t--){
        int n;
        cin >>n;
        int a[n];
        for(int i=0;i<n;i++){
            cin >> a[i];
        }

        Solution ob;
        ob.sort012(a, n);

        for(int i=0;i<n;i++){
            cout << a[i]  << " ";
        }

        cout << endl;
        
        
    }
    return 0;
}


// } Driver Code Ends

File: Difficulty: Easy/Sort 0s, 1s and 2s/sort-0s-1s-and-2s.java
//{ Driver Code Starts
import java.io.*;
import java.util.*;

class GFG {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int t = Integer.parseInt(br.readLine());

        while (t-- > 0) {
            String input = br.readLine();
            String[] inputArray = input.split("\\s+");
            int a[] = new int[inputArray.length];

            for (int i = 0; i < a.length; i++) a[i] = Integer.parseInt(inputArray[i]);

            Solution ob = new Solution();
            ob.sort012(a);

            for (int num : a) {
                System.out.print(num + " ");
            }
            System.out.println();
            System.out.println("~");
        }
    }
}


// } Driver Code Ends
class Solution {
    // Function to sort an array of 0s, 1s, and 2s
    public void sort012(int[] arr) {
        // code here
        
        int i=0;
        int j=0;
        int k=arr.length-1;
        
        while(j<=k){
            
            if(arr[j]==0){
                swap(arr,i,j);
                i++;
                j++;
            }
            else if(arr[j]==2){
                swap(arr,j,k);
                k--;
            }
            else{
                
                j++;
            }
            
            
            
        }
        
    }
    public void swap(int []arr,int i1,int i2){
        int temp=arr[i1];
        arr[i1]=arr[i2];
        arr[i2]=temp;
    }
}

//{ Driver Code Starts.
// } Driver Code Ends